File: back/src/services/categories.ts
import type { Data } from "../data/data.ts";

export function categories(data: Data) {
	return {
		query: async (opts: { userId: string; query?: string }) => {
			const query = opts.query ? opts.query?.trim().toLowerCase() : "";

			return await data.categories.query({
				userId: opts.userId,
				query,
			});
		},
	};
}

export type Categories = ReturnType<typeof categories>;

File: back2/src/services/categories.rs
use crate::data::{Category, Data};

pub async fn query(
    data: &Data,
    user_id: &str,
    search_text: &Option<String>,
) -> anyhow::Result<Vec<Category>> {
    let categories = data.query_categories(user_id, search_text).await?;

    Ok(categories)
}
